store/searchtest/user_layer.go

```diff
@@ -6,674 +6,891 @@ package searchtest
+		Tags: []string{ENGINE_ALL},
+	},
+	{
+		Name: "Should support search all users in a team with username containing a hyphen",
+		Fn:   testSearchUsersInTeamUsernameWithHyphen,
+		Tags: []string{ENGINE_ALL},
+	},
+	{
+		Name: "Should support search all users in a team with username containing a underscore",
+		Fn:   testSearchUsersInTeamUsernameWithUnderscore,
+		Tags: []string{ENGINE_ALL},
+	},
 }
 
 func TestSearchUserStore(t *testing.T, s store.Store, testEngine *SearchTestEngine) {
-	runTestSearch(t, s, testEngine, searchUserStoreTests)
-}
-
-func testSearchDatabaseUserSearch(t *testing.T, s store.Store) {
-	u1 := &model.User{
-		Username:  "jimbo1" + model.NewId(),
-		FirstName: "Tim",
-		LastName:  "Bill",
-		Nickname:  "Rob",
-		Email:     "harold" + model.NewId() + "@simulator.amazonses.com",
-		Roles:     "system_user system_admin",
+	th := &SearchTestHelper{
+		Store: s,
 	}
-	_, err := s.User().Save(u1)
+	err := th.SetupBasicFixtures()
 	require.Nil(t, err)
-	defer func() { require.Nil(t, s.User().PermanentDelete(u1.Id)) }()
+	defer th.CleanFixtures()
+	runTestSearch(t, testEngine, searchUserStoreTests, th)
+}
 
-	u2 := &model.User{
-		Username: "jim-bobby" + model.NewId(),
-		Email:    makeEmail(),
-		Roles:    "system_user",
+func testGetAllUsersInChannelWithEmptyTerm(t *testing.T, th *SearchTestHelper) {
+	options := &model.UserSearchOptions{
+		AllowFullNames: true,
+		Limit:          model.USER_SEARCH_DEFAULT_LIMIT,
 	}
-	_, err = s.User().Save(u2)
+	users, err := th.Store.User().AutocompleteUsersInChannel(th.Team.Id, th.ChannelBasic.Id, "", options)
 	require.Nil(t, err)
-	defer func() { require.Nil(t, s.User().PermanentDelete(u2.Id)) }()
-
-	u3 := &model.User{
-		Username: "jimbo3" + model.NewId(),
-		Email:    makeEmail(),
-		DeleteAt: 1,
-		Roles:    "system_admin",
+	th.User.Sanitize(map[string]bool{})
+	th.User2.Sanitize(map[string]bool{})
+	th.assertUsersMatchInAnyOrder(t, []*model.User{th.User}, users.InChannel)
+	th.assertUsersMatchInAnyOrder(t, []*model.User{th.User2}, users.OutOfChannel)
+}
+func testHonorChannelRestrictionsAutocompletingUsers(t *testing.T, th *SearchTestHelper) {
+	userAlternate, err := th.createUser("user-alternate", "user-alternate", "user", "alternate")
+	require.Nil(t, err)
+	defer th.deleteUser(userAlternate)
+	err = th.addUserToTeams(userAlternate, []string{th.Team.Id})
+	require.Nil(t, err)
+	_, err = th.addUserToChannels(userAlternate, []string{th.ChannelBasic.Id})
+	require.Nil(t, err)
+	options := &model.UserSearchOptions{
+		AllowFullNames:   true,
+		Limit:            model.USER_SEARCH_DEFAULT_LIMIT,
+		ViewRestrictions: &model.ViewUsersRestrictions{Channels: []string{th.ChannelBasic.Id}},
 	}
-	_, err = s.User().Save(u3)
+	t.Run("Autocomplete users with channel restrictions", func(t *testing.T) {
+		users, apperr := th.Store.User().AutocompleteUsersInChannel(th.Team.Id, th.ChannelBasic.Id, "", options)
+		require.Nil(t, apperr)
+		th.User.Sanitize(map[string]bool{})
```

Comment:
Why are we sanitizing the users after running the searches?

Reply:
If we don't sanitize the results and the fixture doesn't match. We can sanitize before but I rather prefer to check for error prior to spend time sanitizing the fixture and the output. You can see the error trace here:
```
Error Trace:	helper.go:426
                	            				user_layer.go:182
                	            				testlib.go:71
                	Error:      	element &{9566t8xdj7dhxggpj3o1qntzah %!s(int64=1586780318282) %!s(int64=1586780318282) %!s(int64=0) basicusername1 $2a$10$Sy8OWemnM.aZSi1RbQ9AyulIHBAQi26SfrR/ds7YNtR7IYkYHCv2. %!s(*string=<nil>)  success_dobpyexxxbytbf9bnq7n6fekzo@simulator.amazon.com %!s(bool=false) basicnickname1 basicfirstname1 basiclastname1   %!s(bool=false) map[] map[channel:true comments:never desktop:mention desktop_sound:true email:true first_name:false mention_keys: push:mention push_status:away] %!s(int64=1586780318282) %!s(int64=0) %!s(int=0) en map[automaticTimezone: manualTimezone: useAutomaticTimezone:true] %!s(bool=false)  %!s(int64=0) %!s(bool=false)  %!s(int64=0)  %!s(int64=0)} appears more times in [%!s(*model.User=&{9566t8xdj7dhxggpj3o1qntzah 1586780318282 1586780318282 0 basicusername1 $2a$10$Sy8OWemnM.aZSi1RbQ9AyulIHBAQi26SfrR/ds7YNtR7IYkYHCv2. <nil>  success_dobpyexxxbytbf9bnq7n6fekzo@simulator.amazon.com false basicnickname1 basicfirstname1 basiclastname1   false map[] map[channel:true comments:never desktop:mention desktop_sound:true email:true first_name:false mention_keys: push:mention push_status:away] 1586780318282 0 0 en map[automaticTimezone: manualTimezone: useAutomaticTimezone:true] false  0 false  0  0})] than in [%!s(*model.User=&{9566t8xdj7dhxggpj3o1qntzah 1586780318282 1586780318282 0 basicusername1  0xc001b446d0  success_dobpyexxxbytbf9bnq7n6fekzo@simulator.amazon.com false basicnickname1 basicfirstname1 basiclastname1   false map[] map[channel:true comments:never desktop:mention desktop_sound:true email:true first_name:false mention_keys: push:mention push_status:away] 1586780318282 0 0 en map[automaticTimezone: manualTimezone: useAutomaticTimezone:true] false  0 false  0  0})]
```

Reply:
Discussed offline, we decided to move the sanitization process to the check helper so it's less verbose at the test level. Resolving!